File: rsyncy.py
#!/usr/bin/env python3

import os
import queue
import re
import select
import subprocess
import sys
import time
import types
from threading import Thread
from datetime import datetime

_re_chk = re.compile(r"(..)-.+=(\d+)/(\d+)")

# start inline from laktakpy


class CLI:
    class style:
        reset = "\033[0m"
        bold = "\033[01m"

    class esc:
        right = "\033[C"

        def clear_line(opt=0):
            # 0=to end, 1=from start, 2=all
            return "\033[" + str(opt) + "K"

    def get_col_bits():
        c, t = os.environ.get("COLORTERM", ""), os.environ.get("TERM", "")
        if c in ["truecolor", "24bit"]:
            return 24
        elif c == "8bit" or "256" in t:
            return 8
        else:
            return 4

    # 4bit system colors
    def fg4(col):
        # black=0,red=1,green=2,orange=3,blue=4,purple=5,cyan=6,lightgrey=7
        # darkgrey=8,lightred=9,lightgreen=10,yellow=11,lightblue=12,pink=13,lightcyan=14
        return f"\033[{(30+col) if col<8 else (90-8+col)}m"

    def bg4(col):
        # black=0,red=1,green=2,orange=3,blue=4,purple=5,cyan=6,lightgrey=7
        return f"\033[{40+col}m"

    # 8bit xterm colors
    def fg8(col):
        return f"\033[38;5;{col}m"

    def bg8(col):
        return f"\033[48;5;{col}m"

    def write(*text):
        for t in text:
            sys.stdout.write(str(t))
        sys.stdout.flush()

    def printline(*text, sep=" ", end="\r\n"):
        CLI.write("\r", sep.join([str(t) for t in text]), CLI.esc.clear_line(), end)

    def get_size():
        # returns {columns, lines}
        if sys.stdout.isatty():
            return os.get_terminal_size()
        else:
            return types.SimpleNamespace(columns=80, lines=40)


# end inline from laktakpy


class Rsyncy:
    def __init__(self, rstyle):
        self.bg = rstyle["bg"]
        self.cdim = rstyle["dim"]
        self.ctext = rstyle["text"]
        self.cbar1 = rstyle["bar1"]
        self.cbar2 = rstyle["bar2"]
        self.cspin = rstyle["spin"]
        self.spinner = rstyle["spinner"]
        self.trans = 0
        self.percent = 0
        self.speed = ""
        self.xfr = ""
        self.chk = ""
        self.chk_finished = False
        self.start = datetime.now()

    def parse_stat(self, line):
        # sample: 6,672,528  96%    1.04MB/s    0:00:06 (xfr#1, to-chk=7/12)
        data = [s for s in line.split(" ") if s]
        if len(data) >= 4:
            self.trans, percent, self.speed, timing, *_ = data
            try:
                self.percent = int(percent.strip("%")) / 100
            except Exception as e:
                print("ERROR - can't parse#1:", line, data, e, sep="\n> ")

        # timing is remaining with 4 args, or elapsed with 6
        if len(data) == 6:
            try:
                xfr, chk = data[4:6]
                self.xfr = xfr.strip(",").split("#")[1]
                if self.xfr:
                    self.xfr = "#" + self.xfr

                m = _re_chk.match(chk)
                if m:
                    self.chk_finished = m[1] == "to"
                    todo = int(m[2])
                    total = int(m[3])
                    done = total - todo
                    self.chk = f"{(done/total if total else 0):2.0%} ({total})"
                else:
                    self.chk = ""

            except Exception as e:
                print("ERROR - can't parse#2:", line, data, e, sep="\n> ")

    def draw_stat(self):
        cols = CLI.get_size().columns
        elapsed = datetime.now() - self.start
        if self.chk_finished:
            spin = ""
        else:
            spin = self.spinner[round(elapsed.total_seconds()) % len(self.spinner)]

        # define status (excl. bar)
        # use \xff as a placeholder for the spinner
        parts = [
            o
            for o in [
                f"{self.trans:>11}",
                f"{self.speed:>14}",
                f"{str(elapsed).split('.')[0]}",
                f"{self.xfr}",
                f"scan {self.chk}\xff",
            ]
            if o
        ]

        # reduce to fit
        plen = lambda: sum(len(s) for s in parts) + len(parts)
        while parts and plen() > cols:
            parts.pop(0)

        # add bar in remaining space
        pc = 0
        rcols = cols - plen()
        if rcols > 12:
            pc_width = min(rcols - 7, 30)
            pc = round(self.percent * pc_width)
            parts.insert(
                0,
                f"{self.cbar1}[{self.cbar2}{'#' * pc}{self.cbar1}{':' * (pc_width-pc)}]{self.ctext}"
                + f"{self.percent:>5.0%}",
            )
            rcols -= pc_width + 7
        elif rcols > 5:
            parts.insert(0, f"{self.percent:>5.0%}")
            rcols -= 5

        # get delimiter size
        delim = f"{self.cdim}|{self.ctext}"
        if rcols > (len(parts) - 1) * 2:
            delim = " " + delim + " "

        # render with delimiter
        status = delim.join(parts).replace("\xff", f"{self.cspin}{spin}{self.ctext}")

        # write and position cursor on bar
        CLI.write(
            "\r",
            self.bg,
            status,
            CLI.esc.clear_line(),
            "\r",
            f"{CLI.esc.right * pc}",
            CLI.style.reset,
        )

    def parse_line(self, line, is_stat):
        line = line.decode().strip(" ")
        if not line:
            return

        is_stat = is_stat or line[0] == "\r"
        line = line.replace("\r", "")

        if is_stat:
            self.parse_stat(line)
            self.draw_stat()
        elif line[-1] == "/":
            # skip directories
            pass
        else:
            CLI.printline(line)
            self.draw_stat()

    def read(self, fd):
        line = b""
        while True:
            stat = select.select([fd], [], [], 0.2)
            if fd in stat[0]:
                ch = os.read(fd, 1)
                if ch == b"":
                    # exit
                    self.parse_line(line, False)
                    break

                elif ch == b"\r":
                    self.parse_line(line, True)
                    line = b"\r"
                elif ch == b"\n":
                    self.parse_line(line, False)
                    line = b""
                else:
                    line += ch

            else:
                # no new input
                if line:
                    # assume this is a status update
                    self.parse_line(line, True)
                    line = b""
                else:
                    # waiting for input
                    self.draw_stat()
                    time.sleep(0.5)

        CLI.printline("")


def run_rsync(args, write_pipe, rc):
    # prefix rsync and add args required for progress
    args = ["rsync"] + args + ["--info=progress2", "--no-v", "-hv"]
    try:
        p = subprocess.Popen(args, stdout=write_pipe)
        p.wait()
    finally:
        os.close(write_pipe)
        rc.put(p.returncode)


if __name__ == "__main__":

    if CLI.get_col_bits() >= 8:
        rstyle = {
            "bg": CLI.bg8(238),
            "dim": CLI.fg8(241),
            "text": CLI.fg8(250),
            "bar1": CLI.fg8(243),
            "bar2": CLI.fg8(43),
            "spin": CLI.fg8(228) + CLI.style.bold,
            "spinner": ["-", "\\", "|", "/"],
        }
    else:
        rstyle = {
            "bg": CLI.bg4(7),
            "dim": CLI.fg4(8),
            "text": CLI.fg4(0),
            "bar1": CLI.fg4(8),
            "bar2": CLI.fg4(0),
            "spin": CLI.fg4(14) + CLI.style.bold,
            "spinner": ["-", "\\", "|", "/"],
        }
    rsyncy = Rsyncy(rstyle)

    try:

        if len(sys.argv) == 1:
            if sys.stdin.isatty():
                print("rsyncy is an rsync wrapper with a progress bar.")
                print(
                    "Please specify your rsync options as you normally would but use rsyncy instead of rsync."
                )
            else:
                # receive pipe from rsync
                rsyncy.read(sys.stdin.fileno())
        else:
            read_pipe, write_pipe = os.pipe()
            rc = queue.Queue()
            t = Thread(target=run_rsync, args=(sys.argv[1:], write_pipe, rc))
            t.start()
            rsyncy.read(read_pipe)
            t.join()
            sys.exit(rc.get())

    except KeyboardInterrupt:
        pass

File: go/rsyncy.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"io"
	"log"
	"os"
	"os/exec"
	"os/signal"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"syscall"
	"time"

	"github.com/laktak/lterm"
	"golang.org/x/term"
)

const (
	escMoveRight = "\033[C"
)

var reChk = regexp.MustCompile(`(..)-.+=(\d+)/(\d+)`)

type Rstyle struct {
	bg      string
	dim     string
	text    string
	bar1    string
	bar2    string
	spin    string
	spinner []string
}

type Rsyncy struct {
	style       Rstyle
	trans       string
	percent     float64
	speed       string
	xfr         string
	chk         string
	chkFinished bool
	start       time.Time
	statusOnly  bool
}

func NewRsyncy(rstyle Rstyle) *Rsyncy {
	return &Rsyncy{
		style: rstyle,
		start: time.Now(),
	}
}

func (r *Rsyncy) parseRsyncStat(line string) bool {
	// sample: 3.93M   5%  128.19kB/s    0:00:29 (xfr#208, ir-chk=2587/2821)
	// sample: 130.95M  29%  207.03kB/s    0:10:17 (xfr#4000, to-chk=1000/5055)
	data := strings.Fields(line)

	if len(data) >= 4 &&
		strings.HasSuffix(data[1], "%") {
		r.trans = data[0]
		if p, err := strconv.ParseFloat(strings.TrimSuffix(data[1], "%"), 64); err == nil {
			r.percent = p / 100.0
		} else {
			// skip
			log.Printf("ERROR - can't parse#1: '%s' in %s\n", data[1], line)
		}
		r.speed = data[2]
		// ignore data[3] (time)

		if len(data) == 6 {
			xfr := strings.Split(strings.TrimSuffix(data[4], ","), "#")
			if len(xfr) == 2 {
				r.xfr = "#" + xfr[1]
			}

			match := reChk.FindStringSubmatch(data[5])
			if len(match) == 4 {
				r.chkFinished = match[1] == "to"
				todo, errTodo := strconv.Atoi(match[2])
				total, errTotal := strconv.Atoi(match[3])
				if errTodo == nil && errTotal == nil && total > 0 {
					done := total - todo
					chkPercent := float64(done) * 100.0 / float64(total)
					r.chk = fmt.Sprintf("%2.0f%% (%d)", chkPercent, total)
				} else {
					r.chk = ""
				}
			} else {
				r.chk = ""
			}
		}
		return true
	}
	return false
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func formatDuration(d time.Duration) string {
	return fmt.Sprintf("%d:%02d:%02d", int(d.Hours()), int(d.Minutes())%60, int(d.Seconds())%60)
}

func (r *Rsyncy) drawStat() {
	cols := lterm.GetWidth()
	elapsed := time.Since(r.start)
	elapsedStr := formatDuration(elapsed)

	spin := ""
	if !r.chkFinished && len(r.style.spinner) > 0 {
		spin = r.style.spinner[int(elapsed.Seconds())%len(r.style.spinner)]
	}

	// define status (excl. bar)
	// use \xff as a placeholder for the spinner
	parts := []string{
		fmt.Sprintf("%11s", r.trans),
		fmt.Sprintf("%14s", r.speed),
		elapsedStr,
		r.xfr,
		fmt.Sprintf("scan %s\xff", r.chk),
	}

	// reduce to fit
	plen := func(parts []string) int {
		sum := 0
		for _, s := range parts {
			sum += len(s)
		}
		return sum + len(parts)
	}

	for len(parts) > 0 && plen(parts) > cols {
		parts = parts[1:]
	}

	// add bar in remaining space
	pc := 0
	rcols := cols - plen(parts)
	pcStr := ""
	if rcols > 12 {
		pcWidth := min(rcols-7, 30)
		pc = int(r.percent * float64(pcWidth))
		pcStr = fmt.Sprintf("%s[%s%s%s%s]%s%4.0f%%",
			r.style.bar1, r.style.bar2, strings.Repeat("#", pc),
			r.style.bar1, strings.Repeat(":", pcWidth-pc),
			r.style.text, r.percent*100,
		)
		rcols -= pcWidth + 7
	} else if rcols > 5 {
		pcStr = fmt.Sprintf("%5.0f%%", r.percent*100)
		rcols -= 5
	}
	if pcStr != "" {
		parts = append([]string{pcStr}, parts...)
	}

	// get delimiter size
	delim := fmt.Sprintf("%s|%s", r.style.dim, r.style.text)
	if rcols > (len(parts)-1)*2 {
		delim = " " + delim + " "
	}

	// render with delimiter
	status := strings.Replace(strings.Join(parts, delim), "\xff", r.style.spin+spin+r.style.text, 1)

	if !r.statusOnly {
		// write and position cursor on bar
		lterm.Write("\r", r.style.bg, status, lterm.ClearLine(0), "\r", strings.Repeat(escMoveRight, pc), lterm.Reset)
	} else {
		lterm.Write("\r\n", r.style.bg, status, lterm.Reset)
	}
}

func (r *Rsyncy) parseLine(lineBytes []byte, isStatHint bool) {
	line := string(bytes.TrimSpace(lineBytes))
	if line == "" {
		return
	}

	isStat := isStatHint || bytes.HasPrefix(lineBytes, []byte("\r"))
	line = strings.TrimPrefix(line, "\r")

	if isStat {
		if !r.parseRsyncStat(line) && !r.statusOnly {
			lterm.Printline("\r", line)
		}
		r.drawStat()
	} else if strings.HasSuffix(line, "/") {
		// skip directories
	} else if !r.statusOnly {
		lterm.Printline("\r", line)
		r.drawStat()
	}
}

func (r *Rsyncy) readOutput(reader io.Reader) {
	bufReader := bufio.NewReader(reader)
	var lineBuffer bytes.Buffer
	ticker := time.NewTicker(200 * time.Millisecond)
	defer ticker.Stop()
	inputChan := make(chan byte, 1024)
	errChan := make(chan error, 1)

	// read bytes and send them to the channel
	go func() {
		for {
			b, err := bufReader.ReadByte()
			if err != nil {
				errChan <- err // signal error (including io.EOF)
				return
			}
			inputChan <- b
		}
	}()

	for {
		select {
		case b := <-inputChan:
			ticker.Reset(200 * time.Millisecond)

			if b == '\r' {
				r.parseLine(lineBuffer.Bytes(), true)
				lineBuffer.Reset()
				lineBuffer.WriteByte('\r')
			} else if b == '\n' {
				r.parseLine(lineBuffer.Bytes(), false)
				lineBuffer.Reset()
			} else {
				lineBuffer.WriteByte(b)
			}

		case <-ticker.C:
			// no new input
			if lineBuffer.Len() > 0 {
				// assume this is a status update
				r.parseLine(lineBuffer.Bytes(), true)
				lineBuffer.Reset()
			} else {
				r.drawStat()
			}

		case <-errChan:
			// exit
			r.parseLine(lineBuffer.Bytes(), false)
			lterm.Printline("\r")
			return
		}
	}
}

func runRsync(args []string, wg *sync.WaitGroup, rcChan chan<- int, stdoutWriter io.WriteCloser) {
	defer wg.Done()
	defer stdoutWriter.Close()

	// prefix rsync and add args required for progress
	args = append(args, "--info=progress2", "--no-v", "-hv")

	cmd := exec.Command("rsync", args...)
	cmd.Stdout = stdoutWriter
	cmd.Stderr = os.Stderr

	err := cmd.Run()
	exitCode := 0
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		} else {
			log.Printf("Error running rsync: %v", err)
			exitCode = 1
		}
	}
	rcChan <- exitCode
}

func main() {
	log.SetFlags(0)

	var rstyle Rstyle
	var spinner = []string{"-", "\\", "|", "/"}
	if lterm.GetTermColorBits() >= 8 {
		rstyle = Rstyle{
			bg:      lterm.Bg8(238),
			dim:     lterm.Fg8(241),
			text:    lterm.Fg8(250),
			bar1:    lterm.Fg8(243),
			bar2:    lterm.Fg8(43),
			spin:    lterm.Fg8(228) + lterm.Bold,
			spinner: spinner,
		}
	} else {
		rstyle = Rstyle{
			bg:      lterm.Bg4(7),
			dim:     lterm.Fg4(8),
			text:    lterm.Fg4(0),
			bar1:    lterm.Fg4(8),
			bar2:    lterm.Fg4(0),
			spin:    lterm.Fg4(14) + lterm.Bold,
			spinner: spinner,
		}
	}

	rsyncy := NewRsyncy(rstyle)

	// status only mode (use with lf)
	if strings.HasSuffix(os.Args[0], "rsyncy-stat") {
		rsyncy.statusOnly = true
	}

	// handle ctrl+c
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		<-sigChan
		lterm.Write(lterm.Reset, "\r", lterm.ClearLine(0), "\r\naborted\r\n")
		os.Exit(1)
	}()

	args := os.Args[1:]

	if len(args) == 0 {
		stdinFd := int(os.Stdin.Fd())
		if term.IsTerminal(stdinFd) {
			fmt.Println("rsyncy is an rsync wrapper with a progress bar.")
			fmt.Println("Please specify your rsync options as you normally would but use rsyncy instead of rsync.")
		} else {
			// receive pipe from rsync
			rsyncy.readOutput(os.Stdin)
		}
	} else {
		readPipe, writePipe, err := os.Pipe()
		if err != nil {
			log.Fatalf("error creating pipe: %v", err)
			os.Exit(1)
		}

		rcChan := make(chan int, 1)
		var wg sync.WaitGroup
		wg.Add(1)

		go runRsync(args, &wg, rcChan, writePipe)
		rsyncy.readOutput(readPipe)
		wg.Wait()
		exitCode := <-rcChan
		os.Exit(exitCode)
	}
}
